File: typescript/src/core/modules/auth/data-model/schema.ts
import joi from "joi";

// Import constants
import { USERNAME_REGEX, PASSWORD_REGEX } from "../../../../utils/constants";

import { toDescriptiveObject } from "../../../validation/joi/helper";

export const signInDataSchema = joi.object({
  username: joi.string().pattern(USERNAME_REGEX).required().messages({
    "string.pattern.base":
      "Username must be 3–30 characters long and only contain letters, numbers, dots (.), underscores (_) or hyphens (-).",
    "string.base": "Username must be a string.",
    "string.empty": "Username cannot be empty.",
    "any.required": "Username is required.",
  }),
  password: joi.string().pattern(PASSWORD_REGEX).required().messages({
    "string.pattern.base":
      "Password must be 8–256 characters long, include at least one uppercase letter, one lowercase letter, one number, and one special character.",
    "string.base": "Password must be a string.",
    "string.empty": "Password cannot be empty.",
    "any.required": "Password is required.",
  }),
});

export const refreshTokensDataSchema = joi.object({
  refreshToken: joi.string().required().messages({
    "string.base": "RefreshToken must be a string",
    "string.empty": "RefreshToken cannot be empty",
    "any.required": "RefreshToken is required",
  }),
});

export const signInDataDescriptiveObject =
  toDescriptiveObject(signInDataSchema);

export const signInResultDescriptiveObject = {
  type: "object",
  properties: {
    auth: {
      type: "object",
      properties: {
        idToken: { type: "string" },
        accessToken: { type: "string" },
        refreshToken: { type: "string" },
        expiresIn: { type: "number" },
      },
    },
  },
};

export const refreshTokensDataDescriptiveObject = toDescriptiveObject(
  refreshTokensDataSchema,
);

export const refreshTokensResultDescriptiveObject = {
  type: "object",
  properties: {
    auth: {
      type: "object",
      properties: {
        idToken: { type: "string" },
        accessToken: { type: "string" },
        refreshToken: { type: "string" },
        expiresIn: { type: "number" },
      },
    },
  },
};

File: python/src/core/modules/auth/data_model/schema.py
from pydantic import BaseModel, Field

# Import constants
from utils.constants import USERNAME_PATTERN, PASSWORD_PATTERN


class SignInDataSchema(BaseModel):
    username: str = Field(..., min_length=3, max_length=50, pattern=USERNAME_PATTERN)
    password: str = Field(..., min_length=8, max_length=128, pattern=PASSWORD_PATTERN)


class RefreshTokensDataSchema(BaseModel):
    refreshToken: str = Field(..., min_length=10)


class AuthTokensSchema(BaseModel):
    idToken: str
    accessToken: str
    refreshToken: str
    expiresIn: int


class RefreshTokensSchema(BaseModel):
    idToken: str
    accessToken: str
    expiresIn: int


class SignInResultSchema(BaseModel):
    auth: AuthTokensSchema


class RefreshTokensResulSchema(BaseModel):
    auth: RefreshTokensSchema


SignInDataDecriptiveObject = SignInDataSchema(
    username="myuserName02", password="MyPassword@#123"
).model_dump()

RefreshTokensDescriptiveObject = RefreshTokensDataSchema(
    refreshToken="eqilaksdjlfkjalsdkjclkjasldkfjlaksdjflkajsdlfkjaslkdfj"
).model_dump()
